File: content/realm/examples/node/legacy/Examples/asymmetric-sync.ts
import Realm, { BSON, ObjectSchema } from "realm";

const app = new Realm.App({ id: "js-flexible-oseso" });

// :snippet-start: asymmetric-sync-object
class WeatherSensor extends Realm.Object<WeatherSensor> {
  _id!: Realm.BSON.ObjectId;
  deviceId!: string;
  temperatureInFahrenheit!: number;
  barometricPressureInHg!: number;
  windSpeedInMph!: number;

  static schema: ObjectSchema = {
    name: "WeatherSensor",
    // sync WeatherSensor objects one way from your device
    // to your Atlas database.
    asymmetric: true,
    primaryKey: "_id",
    properties: {
      _id: "objectId",
      deviceId: "string",
      temperatureInFahrenheit: "int",
      barometricPressureInHg: "float",
      windSpeedInMph: "float",
    },
  };
}
// :snippet-end:

describe("Asymmetric Sync", () => {
  beforeAll(async () => {
    // Close and remove all realms in the default directory.
    Realm.clearTestState();

    const credentials = Realm.Credentials.anonymous();
    await app.logIn(credentials);
  });

  afterAll(async () => {
    app.currentUser?.logOut;
  });

  test("create asymmetric object", async () => {
    if (!app.currentUser) {
      return;
    }

    const weatherSensorPrimaryKey = new BSON.ObjectID();

    // :snippet-start: open-realm
    const realm = await Realm.open({
      schema: [WeatherSensor],
      sync: {
        user: app.currentUser,
        flexible: true,
      },
    });
    // :snippet-end:

    // :snippet-start: write-asymmetric-object
    // :replace-start: {
    //    "terms": {
    //       "weatherSensorPrimaryKey": "new BSON.objectId()"
    //    }
    // }
    realm.write(() => {
      realm.create(WeatherSensor, {
        _id: weatherSensorPrimaryKey,
        deviceId: "WX1278UIT",
        temperatureInFahrenheit: 66.7,
        barometricPressureInHg: 29.65,
        windSpeedInMph: 2,
      });
    });
    // :replace-end:
    // :snippet-end:

    const weatherSensorCollection = await getWeatherSensors();
    const weatherSensor = await weatherSensorCollection.findOne({
      _id: weatherSensorPrimaryKey,
    });

    expect(weatherSensor?._id).toEqual(weatherSensorPrimaryKey);

    // Delete weather sensor documents.
    await weatherSensorCollection.deleteMany({
      deviceId: "WX1278UIT",
    });

    const numberOfWeatherSensorDocuments =
      await weatherSensorCollection.count();

    expect(numberOfWeatherSensorDocuments).toBe(0);

    realm.close();

    function getWeatherSensors(): Promise<
      Realm.Services.MongoDB.MongoDBCollection<WeatherSensor>
    > {
      return new Promise((resolve) => {
        // Wait for weather sensor document to sync, then
        // use mongo client to verify it was created.
        setTimeout(() => {
          const mongodb = app.currentUser!.mongoClient("mongodb-atlas");
          const asyncWeatherSensors = mongodb
            .db("JSFlexibleSyncDB")
            .collection<WeatherSensor>("WeatherSensor");

          resolve(asyncWeatherSensors);
        }, 400);
      });
    }
  });
});

File: content/realm/examples/ios/Examples/AsymmetricSync.swift
// :replace-start: {
//   "terms": {
//     "AsymmetricSync_": "",
//     "FLEX_SYNC_APP_ID": "INSERT_APP_ID_HERE"
//   }
// }

import XCTest
import RealmSwift

// :snippet-start: asymmetric-model
class AsymmetricSync_WeatherSensor: AsymmetricObject {
    @Persisted(primaryKey: true) var _id: ObjectId
    @Persisted var deviceId: String
    @Persisted var temperatureInFahrenheit: Float
    @Persisted var barometricPressureInHg: Float
    @Persisted var windSpeedInMph: Int
    // :remove-start:
    override class func shouldIncludeInDefaultSchema() -> Bool {
        return false
    }
    // :remove-end:
}
// :snippet-end:

class AsymmetricSync: XCTestCase {
    func testAsymmetricSync() async {
        // :snippet-start: connect-and-authenticate
        let app = App(id: FLEX_SYNC_APP_ID)
        do {
            let user = try await login()
            await openSyncedRealm(user: user)
        } catch {
            print("Error logging in: \(error.localizedDescription)")
        }

        func login() async throws -> User {
            let user = try await app.login(credentials: .anonymous)
            return user
        }
        // :snippet-end:

        // :snippet-start: open-asymmetric-sync-realm
        @MainActor
        func openSyncedRealm(user: User) async {
            do {
                var asymmetricConfig = user.flexibleSyncConfiguration()
                asymmetricConfig.objectTypes = [AsymmetricSync_WeatherSensor.self]
                let asymmetricRealm = try await Realm(configuration: asymmetricConfig)
                await useRealm(asymmetricRealm, user)
            } catch {
                print("Error opening realm: \(error.localizedDescription)")
            }
        }
        // :snippet-end:
        // :snippet-start: create-asymmetric-object
        @MainActor
        func useRealm(_ asymmetricRealm: Realm, _ user: User) async {
            try! asymmetricRealm.write {
                asymmetricRealm.create(AsymmetricSync_WeatherSensor.self,
                                       value: [ "_id": ObjectId.generate(),
                                                "deviceId": "WX1278UIT",
                                                "temperatureInFahrenheit": 66.7,
                                                "barometricPressureInHg": 29.65,
                                                "windSpeedInMph": 2
                                                ])
            }
            // :remove-start:
            // Add a delay to give the document time to sync
            sleep(10)
            // Verify the document was added, and then delete it for cleanup.
            let client = app.currentUser!.mongoClient("mongodb-atlas")
            let database = client.database(named: "ios-flexible")
            let collection = database.collection(withName: "AsymmetricSync_WeatherSensor")
            let weatherSensorDocument: Document = ["deviceId": "WX1278UIT"]
            do {
                let foundWeatherSensorDocuments = try await collection.find(filter: weatherSensorDocument)
                print("Found these matching documents: \(foundWeatherSensorDocuments)")
                XCTAssertNotNil(foundWeatherSensorDocuments)
                let deletedResult = try await collection.deleteManyDocuments(filter: weatherSensorDocument)
                print("Deleted documents: \(deletedResult)")
            } catch {
                print("Error finding or deleting documents: \(error.localizedDescription)")
            }
            // :remove-end:
        }
        // :snippet-end:
    }
}
// :replace-end:
